pkg/sql/schemachanger/scplan/internal/rules/current/helpers.go

```diff
@@ -184,36 +189,59 @@ func isIndexDependent(e scpb.Element) bool {
 	return false
 }
 
-// isNonIndexBackedConstraint a non-index-backed constraint.
+// CRDB supports five constraints of two categories:
+// - PK, Unique (index-backed)
+// - Check, UniqueWithoutIndex, FK (non-index-backed)
+func isConstraint(e scpb.Element) bool {
+	return isIndex(e) || isNonIndexBackedConstraint(e)
+}
+
+// isNonIndexBackedConstraint returns true if `e` is a non-index-backed constraint.
 func isNonIndexBackedConstraint(e scpb.Element) bool {
 	switch e.(type) {
-	case *scpb.CheckConstraint, *scpb.ForeignKeyConstraint, *scpb.UniqueWithoutIndexConstraint,
+	case *scpb.CheckConstraint, *scpb.UniqueWithoutIndexConstraint, *scpb.ForeignKeyConstraint,
 		*scpb.ColumnNotNull:
 		return true
+	case *scpb.CheckConstraintNotValid, *scpb.UniqueWithoutIndexConstraintNotValid,
+		*scpb.ForeignKeyConstraintNotValid:
+		return true
 	}
 	return false
 }
 
-func isConstraint(e scpb.Element) bool {
+// isNonIndexBackedComplexConstraint returns true if `e` is a non-index-backed
+// constraint that needs to transition through an intermediate state on its
+// adding/dropping path.
+//
+// We call them "complex" constraint in contrast to simple constraint that
+// does not have an intermediate state.
+func isNonIndexBackedComplexConstraint(e scpb.Element) bool {
 	switch e.(type) {
-	case *scpb.PrimaryIndex, *scpb.SecondaryIndex, *scpb.TemporaryIndex:
-		return true
-	case *scpb.CheckConstraint, *scpb.UniqueWithoutIndexConstraint, *scpb.ForeignKeyConstraint:
+	case *scpb.CheckConstraint, *scpb.UniqueWithoutIndexConstraint, *scpb.ForeignKeyConstraint,
+		*scpb.ColumnNotNull:
 		return true
 	}
 	return false
 }
 
-// isCrossDescriptorConstraint are constraints that might reference
-// other descriptors:
-//   - FKs can reference other table
-//   - Checks can reference other sequences/types
+// isNonIndexBackedCrossDescriptorConstraint returns true if `e` is a
+// non-index-backed constraint and it can potentially reference another
+// descriptor.
+//
+// This filter exists because in general we need to drop the constraint first
+// before dropping referencing/referenced descriptor. Read rules that use
+// this filter for more details.
 //
-// Those constraints need to be dropped first when we are dropping either
-// the referencing or reference descriptor.
-func isCrossDescriptorConstraint(e scpb.Element) bool {
+// TODO (xiang): UniqueWithoutIndex and UniqueWithoutIndexNotValid should
+// also be treated as cross-descriptor constraint because its partial predicate
+// can references other descriptors.
+func isNonIndexBackedCrossDescriptorConstraint(e scpb.Element) bool {
```

Comment:
Like I said in my previous comment, I don't think you need this either. `case *scpb.CheckConstraint, *scpb.UniqueWithoutIndexConstraint, *scpb.ForeignKeyConstraint` should verify `isSubjectTo2VersionInvariant` and the others not.